data_mgr.py

```diff
@@ -0,0 +1,238 @@
+#!/usr/bin/env python3
+
+# -*- coding: utf-8 -*-
+# Copyright (C) 2019 NIWA & British Crown (Met Office) & Contributors.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Create and update the data structure for all workflow services."""
+
+from fnmatch import fnmatchcase
+
+from tornado import gen
+# from google.protobuf.json_format import MessageToDict
+
+# from cylc import flags
+from cylc.exceptions import ClientError, ClientTimeout
+from cylc.ws_messages_pb2 import PbEntireWorkflow
+from cylc.network.scan import MSG_TIMEOUT
+
+
+METHOD_MAP = {
+    'pb_entire_workflow': PbEntireWorkflow
+}
+
+
+async def get_workflow_data(w_id, client, method):
+    # Use already established client
+    try:
+        pb_msg = await client.async_request(method)
+    except ClientTimeout as exc:
+        return (w_id, MSG_TIMEOUT)
+    except ClientError as exc:
+        return (w_id, None)
+    else:
+        ws_data = METHOD_MAP[method]()
+        ws_data.ParseFromString(pb_msg)
+    return (w_id, ws_data)
+
+
+class DataManager(object):
+
+    def __init__(self, ws_mgr):
+        self.ws_mgr = ws_mgr
+        self.data = {}
+
+    # **
+    # Data syncing
+    # **
+    async def entire_workflow_update(self, ids=None):
+        if ids is None:
+            ids = []
+        new_data = {}
+        ws_args = (
+            (w_id, info['req_client'], 'pb_entire_workflow')
+            for w_id, info in self.ws_mgr.workflows.items())
+        items = await gen.multi(
+            get_workflow_data(*args) for args in ws_args
+            if not ids or args[0] in ids)
+        for w_id, result in items:
+            if result is not None and result != MSG_TIMEOUT:
+                new_data[w_id] = result
+        if not ids:
+            # atomic update
+            self.data = new_data
+        else:
+            # apparently faster than update(new_data)
+            for w_id, flow in new_data.items():
+                self.data[w_id] = flow
+
+    # **
+    # Data access filters
+    # **
+    def get_workflow_msgs(self, args):
+        """Return list of workflows."""
+        result = []
+        for flow_msg in self.data.values():
+            if self._workflow_filter(flow_msg, args):
+                result.append(flow_msg)
+        return result
```

Comment:
Equivalent to `return [msg for msg in self.data.values() if self._workflow_filter(msg, args)]` ? Easier to read if we can go up to 86 characters here... otherwise breaking it into another line might still be easier to read?

Reply:
added # TODO  .. there might be further filtering on results.